### PR TITLE
feat: add decorators for api handlers

### PR DESCRIPTION
**Pull request by jfcherng**
Just sharing an idea I got during playing with a ~~mysterious~~ Python language server + `vs_marketplace_client_handler`.

---

With this PR, you can register message handlers with decorators.

For example,

```py
def on_ready(self, api) -> None:
    api.on_notification('eslint/status', self.handle_status)
    api.on_request('eslint/openDoc', self.handle_open_doc)

    # sometimes a handler handles multiple message methods...
    api.on_notification('intelephense/IndexingStarted', self.handle_indexing)
    api.on_notification('intelephense/IndexingEnded', self.handle_indexing)

def handle_status(self, params) -> None:
    ...

def handle_open_doc(self, params, respond) -> None:
    ...

def handle_indexing(self, params) -> None:
    ...
```

can be written as

```py
from lsp_utils import notification_handler, request_handler

@notification_handler("eslint/status")
def handle_status(self, params) -> None:
    ...
    
@request_handler("eslint/openDoc")
def handle_open_doc(self, params, respond) -> None:
    ...

@notification_handler(["intelephense/IndexingStarted", "intelephense/IndexingEnded"])
def handle_indexing(self, params) -> None:
    ...
```

### CODE REVIEW

**Comment by rchl** [comment]
Interesting

Pros:
 - pretty nice and clean syntax
 - avoids the "register" boilerplate

Cons:
 - adds complexity in the code
 - does not necessarily address any issue or annoyance since the current solution is already pretty OK IMO (certainly better than the magic methods in the AbstractPlugin, no offense ;)).

Also, I wonder whether it is types-friendly (haven't tried it myself)? Will it complain if the decorated function doesn't match the expected signature?

**Comment by rchl** [comment]
> ```python
> @as_notification_handler("eslint/status")
> def handle_status(self, params) -> None:
>     pass
> ```

You would have to import the `as_notification_handler` decorator first, right? How would you do that?

**Comment by jfcherng** [comment]
> > ```python
> > @as_notification_handler("eslint/status")
> > def handle_status(self, params) -> None:
> >     pass
> > ```
> 
> You would have to import the `as_notification_handler` decorator first, right? How would you do that?

Ah, yes, the `ApiWrapper` has to be exposed in the `__init__.py` (or put those decorator in something else publicly). Something like

```py

if lsp_version >= (1, 0, 0):  # type: ignore
    # ST 4
    from .vs_marketplace_client_handler_v2 import ApiWrapper
    from .vs_marketplace_client_handler_v2 import VsMarketplaceClientHandler
else:
    # ST 3
    from .vs_marketplace_client_handler import ApiWrapper
    from .vs_marketplace_client_handler import VsMarketplaceClientHandler

__all__ = [
    "ApiWrapper",
    "ApiWrapperInterface",
    "ServerVsMarketplaceResource",
    "VsMarketplaceClientHandler",
]
```

And then you can use it like

```py
from lsp_utils import ApiWrapper

@ApiWrapper.as_notification_handler("eslint/status")
def handle_status(self, params) -> None:
    pass
```

File: st3/lsp_utils/npm_client_handler.py
Comment (jfcherng) [inline st3/lsp_utils/npm_client_handler.py:67]: Should put the registered event_name into a list, otherwise if this decorator is used twice on a same function, only the later one works. But, well, just sharing an idea here :D

**Comment by jfcherng** [comment]
It resolves nothing in practice. Somehow feel it can be useful in some ways (not necessary in this PR).

**Comment by rchl** [comment]
> Ah, yes, the `ApiWrapper` has to be exposed in the `__init__.py` (or put those decorators something else publicly). Something like

Conceptually that would be wrong as it's an implementation-specific class that is not supposed to be exposed.

File: st3/lsp_utils/npm_client_handler.py
Comment (rchl) [inline st3/lsp_utils/npm_client_handler.py:225]: I suppose you wouldn't want to decorate the same function with both decorators as that wouldn't work. So should break early.

**Comment by jfcherng** [comment]
> Also, I wonder whether it is types-friendly (haven't tried it myself)? Will it complain if the decorated function doesn't match the expected signature?

On my side, LSP-py* complains after I make the function signature clear.

```py
T_NOTIFICATION_HANDLER = Callable[["VsMarketplaceClientHandler", Dict[str, Any]], None]
```

and

```py
    @staticmethod
    def as_notification_handler(event_name: str) -> Callable[[T_NOTIFICATION_HANDLER], T_NOTIFICATION_HANDLER]:
        """ Marks the decorated function as a handler for the notification event. """

        def decorator(func: T_NOTIFICATION_HANDLER) -> T_NOTIFICATION_HANDLER:
            setattr(func, ApiWrapper.HANDLER_MARK_NOTIFICATION, event_name)
            return func

        return decorator
```

**Comment by jfcherng** [comment]
> > Ah, yes, the `ApiWrapper` has to be exposed in the `__init__.py` (or put those decorators something else publicly). Something like
> 
> Conceptually that would be wrong as it's an implementation-specific class that is not supposed to be exposed.

Or probably put decorators under `NpmClientHandler`, and use `@self.as_notification_handler(...)`.

File: st3/lsp_utils/__init__.py
Comment (rchl) [inline st3/lsp_utils/__init__.py:2]: I'm now used to the convention where each import is separate (apart from types).

File: st3/lsp_utils/_client_handler/decorator.py
Comment (rchl) [inline st3/lsp_utils/_client_handler/decorator.py:19]: I think the `as_` prefix is redundant. A decorator like `@notification_handler` is as clear and probably more conventional (built-in python decorators like `@classmethod` don't use the `as_` prefix).

File: st3/lsp_utils/generic_client_handler.py
Comment (rchl) [inline st3/lsp_utils/generic_client_handler.py]: `_register_decorated_handlers` maybe?

File: st3/lsp_utils/generic_client_handler.py
Comment (rchl) [inline st3/lsp_utils/generic_client_handler.py]: Does this allow for both decorators to be applied to the same function? That wouldn't make sense as the function signatures differ.

**Review by rchl — COMMENTED** [review]
You should avoid force-push when there are reviewers involved since that makes it harder to see what has changed.

In the end, when ready to merge, there is an option to squash and merge so the history will get cleaned up.

File: st3/lsp_utils/_client_handler/language_handler.py
Comment (rchl) [inline st3/lsp_utils/_client_handler/language_handler.py:152]: Does it need to be done separately for each handler implementation. Did the previous approach of changing generic client handler didn't work?

File: st3/lsp_utils/_client_handler/decorator.py
Comment (jfcherng) [reply st3/lsp_utils/_client_handler/decorator.py:19]: I am fine with whatever names :)

File: st3/lsp_utils/_client_handler/language_handler.py
Comment (jfcherng) [reply st3/lsp_utils/_client_handler/language_handler.py:152]: It works but potentially error-prone if people forget to do `super().on_ready()` in their custom `on_ready()` then `_register_custom_server_event_handlers()` wont be executed.

File: st3/lsp_utils/generic_client_handler.py
Comment (jfcherng) [reply st3/lsp_utils/generic_client_handler.py]: yeah... at this moment I just assume those two are all `Callable[[Any, Any], None]`...

File: st3/lsp_utils/_client_handler/language_handler.py
Comment (rchl) [reply st3/lsp_utils/_client_handler/language_handler.py:152]: But on the other-hand such duplication is worse. The big goal of the refactor was to get rid of as much duplication as possible.

You can add some internal functions to the client interface that will be called by handler implementations before `on_ready` and which generic handler can decorate from.

File: st3/lsp_utils/generic_client_handler.py
Comment (rchl) [reply st3/lsp_utils/generic_client_handler.py]: I would really like the types to be correct and checked so that when decorating with `@notification_handler`, for example, would raise a type error if the function doesn't have a correct number of arguments (and maybe also correct types).

I'm not certain if it's possible for decorators to affect type checking though...

Comment (rchl) [reply st3/lsp_utils/generic_client_handler.py]: > yeah... at this moment I just assume those two are all `Callable[[Any, Any], None]`...

Types being incorrect is one thing but I guess it would also crash at runtime since the number of arguments wouldn't match the passed number of arguments.

File: st3/lsp_utils/_client_handler/language_handler.py
Comment (jfcherng) [reply st3/lsp_utils/_client_handler/language_handler.py]: I guess https://github.com/sublimelsp/lsp_utils/pull/38/commits/b9f81229128f1780ba2c29de4ff56175abfd9f6b resolves this at a certain degree.

File: st3/lsp_utils/npm_client_handler.py
Comment (jfcherng) [reply st3/lsp_utils/npm_client_handler.py]: Yes, I think it doesn't make sense to have a handler handles both notification and request. Will change.

File: st3/lsp_utils/_client_handler/decorator.py
Comment (rchl) [inline st3/lsp_utils/_client_handler/decorator.py:22]: I don't understand how such a decorator can work since the argument passed to decoration is the original function, I believe, not whatever `server_events` is.

File: st3/lsp_utils/npm_client_handler.py
Comment (jfcherng) [reply st3/lsp_utils/npm_client_handler.py]: Resolved as of https://github.com/sublimelsp/lsp_utils/pull/38/commits/179920b79cf7cdac68ffd5ab5cc5d68203f3e7da

File: st3/lsp_utils/_client_handler/decorator.py
Comment (jfcherng) [reply st3/lsp_utils/_client_handler/decorator.py:14]: I have no special opinion about this just because the codes can run with an Iterable. Limit it to be a list is okay IMO.

Comment (jfcherng) [reply st3/lsp_utils/_client_handler/decorator.py]: Thanks! I was actually out of words for naming :rofl: ... Naming is harder than cache invalidation.

File: st3/lsp_utils/generic_client_handler.py
Comment (jfcherng) [reply st3/lsp_utils/generic_client_handler.py]: Will investigate (my) tomorrow!

File: st3/lsp_utils/_client_handler/decorator.py
Comment (rchl) [inline st3/lsp_utils/_client_handler/decorator.py:75]: I think you can just `break` here instead of using `is_registered` variable for that purpose.

Comment (jfcherng) [reply st3/lsp_utils/_client_handler/decorator.py]: Nice catch!

File: st3/lsp_utils/_client_handler/decorator.py
Comment (rchl) [inline st3/lsp_utils/_client_handler/decorator.py:14]: Types typically use CamelCase convention.
Constants (like `_HANDLER_MARKS` here) use UPPERCASE.

File: st3/lsp_utils/generic_client_handler.py
Comment (rchl) [reply st3/lsp_utils/generic_client_handler.py]: If you enable pushes to this PR then I'll push a fix.

Comment (jfcherng) [reply st3/lsp_utils/generic_client_handler.py]: image

Weird. I remember there used to be a "allow edit by maintainer" in the sidebar but there isn't.

Comment (rchl) [reply st3/lsp_utils/generic_client_handler.py]: Maybe it's on the bottom, where the comment box is?

Comment (jfcherng) [reply st3/lsp_utils/generic_client_handler.py]: Hmmm... can't find it

**Comment by rchl** [comment]
I'd also like this to be documented in new docs.
I can handle that.

**Comment by jfcherng** [comment]
MUCH appreciated for PR reviewing (many lessons learned) and documenting. And that lsp_utils refactoring :heart:

File: st3/lsp_utils/generic_client_handler.py
Comment (jfcherng) [reply st3/lsp_utils/generic_client_handler.py]: Hey seriously I can't find it in https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork either now or while creating a PR...

**Comment by jfcherng** [comment]
The link to documents in the README is wrong btw.

File: st3/lsp_utils/_client_handler/decorator.py
Comment (jfcherng) [inline st3/lsp_utils/_client_handler/decorator.py:3]: There is no `TypeVar` in the master branch of LSP.
https://github.com/sublimelsp/LSP/blob/master/plugin/core/typing.py

Comment (rchl) [reply st3/lsp_utils/_client_handler/decorator.py:3]: Good point. Will add it there.

File: st3/lsp_utils/_client_handler/decorator.py
Comment (jfcherng) [inline st3/lsp_utils/_client_handler/decorator.py:13]: We can annotate the first argument (`self`) as `ClientHandlerInterface`.

**Comment by predragnikolic** [comment]
What do you think of a shorter name for the decorators?

`@request_handler` -> `@on_request` -> `@request`
`@notification_handler` -> `@on_notification` -> `@notification`

**Comment by rchl** [comment]
> What do you think of a shorter name for the decorators?
> 
> `@request_handler` -> `@on_request`
> `@notification_handler` -> `@on_notification`

I feel that the existing naming is more in line with conventions I'm familiar with where decorator describes what it turns the function into (like `@classmethod`). `on_` prefix would kinda suggest that it listens to something on the function it decorates.

**Comment by predragnikolic** [comment]
> where decorator describes what it turns the function into

When you put it on that way it makes sense to name it `request_handler` `notification_handler`.

Here is the reason on why I suggested the name change.
Because the `@notification_handler` does the same thing the same thing `api.on_notification`, I felt like that they should be named the same. 

```
def on_ready(self, api) -> None:
    api.on_notification('eslint/status', self.handle_status)

def handle_status(self, params) -> None:
    ...

# the above achieves the same result as the bellow

@on_notification("eslint/status")
def handle_status(self, params) -> None:
    ...
```

the @on_notification decorator for me would suggest that the function it decorates will be executed when the notification is received.

**Comment by rchl** [comment]
I said:

> `on_` prefix would kinda suggest that it listens to something on the function it decorates.

and I think the example you made:

> api.on_notification('eslint/status', self.handle_status)

is also in line with that statement. In the case of a central `api` object, we are adding a listener to that object so I think the names being different in those cases is appropriate. It's a different way of setting up a handler.

File: st3/lsp_utils/_client_handler/decorator.py
Comment (rchl) [reply st3/lsp_utils/_client_handler/decorator.py:13]: That doesn't work and I don't have enough resources to figure out why is that.

Possibly due to covariance/contravariance thing.

**Review by rchl — APPROVED** [review]

**Comment by jfcherng** [comment]
Can't wait for the next release. So many improvements for LSP-* devs :D

**Comment by jfcherng** [comment]
Currently the decorator methods can't handle some special notifications like `$/progress`, `window/logMessage`. Maybe we need to refactor the implementation of this PR, to make use of https://github.com/sublimelsp/LSP/issues/2494.

**Comment by rchl** [comment]
I feel like it was really only meant to handle custom requests/notifications. I think that it's consistent in that case?

It would be a bit ambiguous to specify what should happen if both LSP and a plugin handle given request at least.

**Comment by jfcherng** [comment]
> It would be a bit ambiguous to specify what should happen if both LSP and a plugin handle given request at least.

If the client wants to make it into a mess, its it's issue.

**Pull request merged**
